File: Beta estimation.R
#----------------------------------------
# Beta model estimation (SPF distrib.)
# in R (16.01.2023)
#----------------------------------------

#---------------------------------
# 1. Method of Moments estimation
#---------------------------------

set.seed(2023)
# one realization of a Beta(1/theta,1) with theta = 3
theta = 3
rbeta(n = 1, shape1 = 1/theta, shape2 = 1) 
# [1] 0.3323928

# on average, we should get about 1/4 since for theta = 3
mean(rbeta(n = 100000, shape1 = 1/theta, shape2 = 1) )
# [1] 0.2494053

# plotting
seq = seq(0, 10, length=1000)
par(mfrow = c(1,2))
plot(seq, dbeta(seq, 1/3, 1), type='l', col = 'red', lwd = 2, main = 'Beta(1/3,1)', xlim = c(0,4))
hist(rbeta(n = 10000, shape1 = 1/3, shape2 = 1) , col = 'red', breaks = 25, main = 'Sample of size 10,000',
     freq = NULL, border = 'white', ylab = 'count', xlab = 'I_hat')

# Method of Moments estimator
MoMBeta <- function(x) {
  
  n <- length(x)
  sample_moment <- sum(x) / n
  
  theta_mom <- (1 / sample_moment) -1
  alpha_mom <- 1 / theta_mom
  
  output <- NULL
  output$alpha_mom <- alpha_mom
  output$theta_mom <- theta_mom
  
  return(output)
}

# generate artificial data, sample of size 100,000
set.seed(2021)
x <- rbeta(n = 100000, shape1 = 1/3, shape2 = 1)

# apply MoMBeta()
MomBeta(x = x)

# $alpha_mom
# [1] 0.3369428

# $theta_mom
# [1] 2.967863

#---------------------------------
# 2. Maximum Likelihood estimation
#---------------------------------

# generate a random sample of n = 5000 from an Beta distribution
set.seed(2023)
n = 5000 ; theta = 3
xi <- rbeta(n = n, shape1 = 1/theta, shape2 = 1) 
head(xi)
# [1] 0.332392794 0.448662938 0.001354903 0.002210476 0.064676188 0.021291173

# Closed-form MLE
theta_hat_formula =  1 / (- n / sum(log(xi))  )
theta_hat_formula 
# [1] 3.023691

# Numerical approximation of the MLE 
mle = optimize(function(theta){sum(dbeta(x = xi, shape1 = 1/theta, shape2 = 1, log = TRUE))},
               interval = c(0, 10),
               maximum = TRUE,
               tol = .Machine$double.eps^0.5)

theta_hat = mle$maximum
theta_hat 
# [1] 3.023691

# Plot the Log-Likelihood function
library(ggplot2)
theta = 3
possible.theta <- seq(0, 10, by = 0.01)

qplot(possible.theta,
      sapply(possible.theta, function(theta) {sum(dbeta(x = xi, shape1 = 1/theta, shape2 = 1, log = TRUE))}),
      geom = 'line',
      ylim = c(-30000, 6000),
      xlab = 'theta',
      ylab = 'Log-Likelihood') +
  geom_vline(xintercept = theta_hat, color = 'red', size=1.5) +
  labs(title = 'Beta Log-Likelihood function (function of theta)',
       subtitle = "Maximum is reached at theta = 3.023691",
       caption = "Artificial dataset of size 5,000") +
  geom_line(size = 1.1) + 
  theme(axis.text = element_text(size = 8),
        axis.title = element_text(size = 10),
        plot.subtitle = element_text(size = 9, face="italic", color="darkred"),
        panel.background = element_rect(fill = "white", colour = "grey50"),
        panel.grid.major = element_line(colour = "grey90"))

#----
# end
#----

File: Beta estimation.py
#----------------------------------------
# Beta model estimation (SPF distrib.)
# in Python (16.01.2023)
#----------------------------------------

#---------------------------------
# 1. Method of Moments estimation
#---------------------------------

import numpy as np
np.random.seed(2023)

from scipy.stats import beta
# one realization of a Beta(1/theta, 1) with theta = 3
theta = 3
beta.rvs(a = 1/theta, size = 1, b = 1)
# array([0.03613599])

import statistics
# on average, we should get one.
statistics.mean(beta.rvs(a = 1/theta, size = 100000, b = 1))
# 0.24922457399140857

# Method of Moments estimator
def MoMBeta(x):
  
  n = len(x)
  sample_moment = np.sum(x) / n
  
  theta_mom = (1 / sample_moment) -1
  alpha_mom = 1 / theta_mom
  
  return theta_mom, alpha_mom

# generate artificial data, sample of size 10,000
np.random.seed(2023)
x = beta.rvs(a = 1/3, size = 100000, b = 1)

# apply MoMgamma()
MoMBeta(x = x)
# (2.9968621974937864, 0.3336823430974835)

#---------------------------------
# 2. Maximum Likelihood estimation
#---------------------------------

# generate a sample of size 5000 from a Beta distribution
np.random.seed(2023)
n = 5000
theta = 3 # true value of the parameter, that we wish to estimate
xi = beta.rvs(a = 1/theta, size = 5000, b = 1)
print(xi)
# [0.03613599 0.61631475 0.00599853 ... 0.23542418 0.00548306 0.82538336]

# Closed-form MLE
theta_hat_formula = 1 / ( - (n / np.sum(np.log(xi))) )  
theta_hat_formula 
# [1] 3.0272294817747585

#T numerical optimization
from scipy import stats

def llikelihood(Theta):  
    # log-likelihood function
    ll = -np.sum(stats.beta.logpdf(xi, a = 1/Theta, b = 1))
    return ll

from scipy.optimize import minimize

# Numerical approximation of the MLE using minimize()
mle = minimize(llikelihood, 
                   x0 = 4, 
                   method = 'BFGS')
print(mle.x)
# [3.02722947]

# Plot the Log-Likelihood function
import matplotlib.pyplot as plt

# set a range of possible parameter values
possibletheta = np.linspace(start = 0, stop = 10, num = 5000)

# compute the log-likelihood function for all possible parameter values
LL = []
for L in possibletheta:
    loglikelihood = np.log(L)- (L * xi)
    LL.append(loglikelihood.sum())

# plotting using matplotlib
px = 1/plt.rcParams['figure.dpi'] 
plt.figure(figsize=(850*px, 400*px))
plt.plot(possibletheta, LL, color = 'black')
plt.axvline(x = mle.x,  color = 'red')
plt.ylim([-30000, 6000])
plt.xlabel('lambda')
plt.ylabel('Log-Likelihood')
plt.title('Log-Likelihood (function of theta)', fontsize = 15, y = 1.1, loc = 'left')
plt.suptitle('Maximum is reached at theta = 3.02722947', y = 0.95, color = 'darkred', x = 0.283)
caption = 'artificial dataset of size 5,000'
plt.text(6.5,-38000, caption, fontsize = 8)
plt.grid(color = 'whitesmoke', linestyle = '-', linewidth = 1.2)

#----
# end
#----
